Add Action Client

jacobperron opened the pull request:
Connects to #257 

* Added `_rclpy_action` extension module for interfacing `rcl_action` library and `rclpy`
* Implemented ActionClient and integrated with Node as a Waitable
* Added unit tests for ActionClient and ClientGoalHandle

I still need to go through and update the code documentation and perhaps add some more unit tests.

jacobperron commented:
CI:
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=6070)](http://ci.ros2.org/job/ci_linux/6070/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=2550)](http://ci.ros2.org/job/ci_linux-aarch64/2550/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=4999)](http://ci.ros2.org/job/ci_osx/4999/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=5926)](http://ci.ros2.org/job/ci_windows/5926/)

sloretz commented on rclpy/test/test_action_client.py:
```diff
@@ -0,0 +1,304 @@
+        cls.executor = SingleThreadedExecutor(context=cls.context)
+        cls.node = rclpy.create_node('TestActionClient', context=cls.context)
+        cls.mock_action_server = MockActionServer(cls.node)
+        cls.feedback = None
```
IIUC this sets feedback to `None` once before all tests are run, but the tests expect `feedback` cleared before each test is run. That can be done with a `setUp()` method instead.

sloretz commented on rclpy/test/test_action_client.py:
```diff
@@ -0,0 +1,304 @@
+        ac = ActionClient(self.node, Fibonacci, 'not_fibonacci')
+        try:
+            future = ac.send_goal_async(Fibonacci.Goal())
+            self.assertFalse(rclpy.spin_once(self.node, executor=self.executor, timeout_sec=2.0))
```
`rclpy.spin_once()` always returns `None`, so the `assertFalse()` is unnecessary.

https://github.com/ros2/rclpy/blob/91996411deba25b662aaef08440a3103ed38c60c/rclpy/rclpy/__init__.py#L92-L93

sloretz replied:
Also since this is being run with `pytest,` plain `assert` statements can be used instead of `unittest`'s `assertFalse()`. `pytest` does a pretty good job printing values from `assert` https://docs.pytest.org/en/latest/example/reportingdemo.html#tbreportdemo

No change required, it's up to you what style you prefer.

sloretz replied:
Since there could be multiple things `spin_once` could spin on, use `timed_spin()` instead?

jacobperron replied:
Updated to use `timed_spin()` 37c98d6

---

I went with `unittest`'s style for consistency with the other tests I looked at in the package, but in retrospect there are a few that use plain `assert` as well. IMO I prefer the aesthetic of plain `assert`. What do you think about an overhaul of all the tests so they're consistent in a follow-up PR?

sloretz replied:
Converting the tests from unittest to pytest style in a follow up PR sounds good to me

jacobperron commented:
I still have to implement the synchronous methods (meant to do that prior to this PR).

sloretz commented on rclpy/src/rclpy/_rclpy_action.c:
```diff
@@ -0,0 +1,885 @@
+    Py_DECREF(pymetaclass); \
+    return NULL; \
+  } \
+  convert_to_py_signature * convert_to_py = get_capsule_pointer(pymetaclass, "_CONVERT_TO_PY"); \
```
Recommend getting all the capsule pointers at the top so `Py_DECREF(pymetaclass)` can be in fewer error blocks.

sloretz commented on rclpy/src/rclpy/impl/common.h:
```diff
@@ -0,0 +1,60 @@
+static PyObject *
+rclpy_convert_to_py_qos_policy(void * profile)
+{
+  PyObject * pyqos_module = PyImport_ImportModule("rclpy.qos");
```
I realize this was copied from _rclpy, just noting this function is not decrementing references or checking for errors like it should.

jacobperron replied:
Addressed in 9870847

jacobperron commented on rclpy/src/rclpy/_rclpy_action.c:
```diff
@@ -0,0 +1,885 @@
+    return NULL;
+  }
+
+  char * action_name = (char *)PyUnicode_1BYTE_DATA(pyaction_name);
```
Using `PyUnicode_AsUTF8()` sounds reasonable to me.

I copied this line from `_rclpy.c`, where it is used throughout, e.g.

https://github.com/ros2/rclpy/blob/0fafaec6b3764654413d0ce37839d7c7178245af/rclpy/src/rclpy/_rclpy.c#L816

Perhaps it's worth making a note to update the `_rclpy.c` in a follow-up PR.

sloretz replied:
Created issue #266 to fix the remaining places in `_rclpy.c`

jacobperron commented on rclpy/src/rclpy/_rclpy_action.c:
```diff
@@ -0,0 +1,885 @@
+}
+
+#define OPTIONS_COPY_QOS_PROFILE(Options, Profile) \
+  if (PyCapsule_IsValid(py ## Profile, "rmw_qos_profile_t")) { \
```
Wow, in doing so I think we've uncovered a bug. It would appear that a capsule for `rmw_qos_profile_t` is never created. Instead `rclpy_convert_to_py_qos_policy()` created a regular Python object of type `QosProfile`. The C struct is not stored (or reconstructed) anywhere.

https://github.com/ros2/rclpy/blob/523c403c60c12dc304564f2e681bf1d6dd637b9d/rclpy/src/rclpy/impl/common.h#L27-L47

Correct me if I'm wrong, but I think this means that all of the checks `_rclpy.c` are failing silently and QoS settings are never copied to the `rcl` layer:

https://github.com/ros2/rclpy/blob/0fafaec6b3764654413d0ce37839d7c7178245af/rclpy/src/rclpy/_rclpy.c#L1179

https://github.com/ros2/rclpy/blob/0fafaec6b3764654413d0ce37839d7c7178245af/rclpy/src/rclpy/_rclpy.c#L1689

https://github.com/ros2/rclpy/blob/0fafaec6b3764654413d0ce37839d7c7178245af/rclpy/src/rclpy/_rclpy.c#L1779

https://github.com/ros2/rclpy/blob/0fafaec6b3764654413d0ce37839d7c7178245af/rclpy/src/rclpy/_rclpy.c#L1936

jacobperron replied:
@clalancette Is there an issue opened somewhere? If not, I can create one.

clalancette replied:
No, I hadn't opened one yet.  Thanks for looking at this!

jacobperron replied:
False alarm. I missed the call to [`rclpy_convert_from_py_qos_policy()`](https://github.com/ros2/rclpy/blob/0fafaec6b3764654413d0ce37839d7c7178245af/rclpy/src/rclpy/_rclpy.c#L3401) in `qos.py`:

https://github.com/ros2/rclpy/blob/0fafaec6b3764654413d0ce37839d7c7178245af/rclpy/rclpy/qos.py#L123-L130

@clalancette This is not the bug you're looking for :robot:

jacobperron commented:
Rebased on master.

@sloretz Thanks for the review so far!

I believe I've address most of your comments in 7ce5e74

jacobperron commented on rclpy/src/rclpy/_rclpy_action.c:
```diff
@@ -0,0 +1,885 @@
+    return NULL;
+  }
+
+  return PyCapsule_New(action_client, "rcl_action_client_t", NULL);
```
I've added a `__del__` method to `ActionClient` to ensure the object is destroyed during garbage collection.

I can add a PyCapsule destructor. But I guess we should choose one strategy for destruction to avoid a double-free?

jacobperron replied:
Another issue that came up is that we should ensure that the `Node` is destroyed *after* the action client, since the rcl finalize functions require a reference to the node. What do you think about adding a `destroy()` method to the `Waitable` interface for `Node` to call or a `Destroyable` interface? I'm not sure if we have to worry about ABI/API issues with the former.

jacobperron replied:
I've added a check in the destroy method to avoid a potential segfault if the action client is destroyed after the node: f2b238c

sloretz replied:
> Another issue that came up is that we should ensure that the Node is destroyed after the action client

There are a bunch of memory issues with nodes and node entities in `rclpy`. What I'd like to see is too big for this PR (if something needs a node, it should hold a reference to the node rather than the other way around), so I'm fine with setting the capsule pointer to a sentinel value `PyNone` for now.

jacobperron commented on rclpy/src/rclpy/_rclpy_action.c:
```diff
@@ -0,0 +1,885 @@
+{
+  const char * pyrmw_profile;
+  if (!PyArg_ParseTuple(
+      args, "z", &pyrmw_profile))
```
Changed to `"s"`

jacobperron commented on rclpy/src/rclpy/_rclpy_action.c:
```diff
@@ -0,0 +1,885 @@
+    return NULL; \
+  } \
+  PyObject * pyrequest_type = PyObject_GetAttrString(pyrequest, "__class__"); \
+  assert(pyrequest_type != NULL); \
```
I think we should also go through `_rclpy.c` and remove `assert`'s in similar places.
I'll create a follow up task for it, if you don't mind.

sloretz commented on rclpy/src/rclpy/_rclpy_action.c:
```diff
@@ -0,0 +1,929 @@
+static PyObject *
+rclpy_action_get_rmw_qos_profile(PyObject * Py_UNUSED(self), PyObject * args)
+{
+  const char * pyrmw_profile;
```
I would remove the `py` prefix since it's not a `PyObject`. There isn't a rule for this in the style guide AFAIK though

jacobperron replied:
Done 34db026

sloretz commented on rclpy/src/rclpy/_rclpy_action.c:
```diff
@@ -0,0 +1,929 @@
+      PyLong_FromSize_t(num_guard_conditions),
+      PyLong_FromSize_t(num_timers),
+      PyLong_FromSize_t(num_clients),
+      PyLong_FromSize_t(num_services));
```
Memory is leaked if `PyLong_FromSize_t()` succeeds but `PyTuple_Pack()` fails.

jacobperron replied:
Fixed  34db026

sloretz commented on rclpy/src/rclpy/_rclpy_action.c:
```diff
@@ -0,0 +1,929 @@
+      PyLong_FromSize_t(num_clients),
+      PyLong_FromSize_t(num_services));
+  if (!result_tuple) {
+    return PyErr_Format(PyExc_RuntimeError, "Failed to create new tuple object");
```
return `NULL` since `PyTuple_Pack` already raised an exception.

jacobperron replied:
Done 34db026

sloretz commented on rclpy/src/rclpy/_rclpy_action.c:
```diff
@@ -0,0 +1,929 @@
+        PyBool_FromLong(is_status_ready),
+        PyBool_FromLong(is_goal_response_ready),
+        PyBool_FromLong(is_cancel_response_ready),
+        PyBool_FromLong(is_result_response_ready));
```
Memory is leaked if `PyBool_FromLong` succeeds but `PyTuple_Pack` fails

jacobperron replied:
Fixed 34db026

sloretz commented on rclpy/src/rclpy/_rclpy_action.c:
```diff
@@ -0,0 +1,929 @@
+        PyBool_FromLong(is_cancel_response_ready),
+        PyBool_FromLong(is_result_response_ready));
+    if (!result_tuple) {
+      return PyErr_Format(PyExc_RuntimeError, "Failed to create result tuple for action client");
```
return `NULL` since `PyTuple_Pack` already set an exception

jacobperron replied:
Done 34db026

sloretz commented on rclpy/src/rclpy/_rclpy_action.c:
```diff
@@ -0,0 +1,929 @@
+        PyBool_FromLong(is_goal_request_ready),
+        PyBool_FromLong(is_cancel_request_ready),
+        PyBool_FromLong(is_result_request_ready),
+        PyBool_FromLong(is_goal_expired));
```
memory is leaked here too if `PyTuple_Pack` fails

jacobperron replied:
Fixed 34db026

sloretz commented on rclpy/src/rclpy/_rclpy_action.c:
```diff
@@ -0,0 +1,929 @@
+        PyBool_FromLong(is_result_request_ready),
+        PyBool_FromLong(is_goal_expired));
+    if (!result_tuple) {
+      return PyErr_Format(PyExc_RuntimeError, "Failed to create result tuple for action server");
```
Also return `NULL`

jacobperron replied:
Done34db026

sloretz commented on rclpy/src/rclpy/_rclpy_action.c:
```diff
@@ -0,0 +1,929 @@
+    return NULL;
+  }
+
+  if (!PyUnicode_Check(pyaction_name)) {
```
No need for this check, `PyUnicode_AsUTF8` [calls `PyUnicode_Check`](https://github.com/python/cpython/blob/ea446409cd5f1364beafd5e5255da6799993f285/Objects/unicodeobject.c#L3870)

jacobperron replied:
Removed 34db026

sloretz commented on rclpy/src/rclpy/_rclpy_action.c:
```diff
@@ -0,0 +1,929 @@
+
+  PyObject * pymetaclass = PyObject_GetAttrString(pyaction_type, "__class__");
+  if (!pymetaclass) {
+    return PyErr_Format(PyExc_AttributeError, "Action type has no '__class__' attribute");
```
return `NULL` since `PyObject_GetAttrString` already raised an exception

jacobperron replied:
Done 34db026

sloretz commented on rclpy/src/rclpy/_rclpy_action.c:
```diff
@@ -0,0 +1,929 @@
+  PyObject * pyts = PyObject_GetAttrString(pymetaclass, "_TYPE_SUPPORT");
+  Py_DECREF(pymetaclass);
+  if (!pyts) {
+    return PyErr_Format(PyExc_AttributeError, "Action type has no '_TYPE_SUPPORT' attribute");
```
return `NULL` since `PyObject_GetAttrString` already raised an exception

jacobperron replied:
Done 34db026

sloretz commented on rclpy/src/rclpy/_rclpy_action.c:
```diff
@@ -0,0 +1,929 @@
+  } \
+  PyObject * pyrequest_type = PyObject_GetAttrString(pyrequest, "__class__"); \
+  if (!pyrequest_type) { \
+    return PyErr_Format(PyExc_AttributeError, \
```
Return `NULL` since `PyObject_GetAttrString` already raised an exception

jacobperron replied:
Done 34db026

sloretz commented on rclpy/src/rclpy/_rclpy_action.c:
```diff
@@ -0,0 +1,929 @@
+  PyObject * pymetaclass = PyObject_GetAttrString(pyrequest_type, "__class__"); \
+  Py_DECREF(pyrequest_type); \
+  if (!pymetaclass) { \
+    return PyErr_Format(PyExc_AttributeError, \
```
Return `NULL` since `PyObject_GetAttrString` already raised an exception

sloretz commented on rclpy/src/rclpy/_rclpy_action.c:
```diff
@@ -0,0 +1,929 @@
+  } \
+  PyObject * pymetaclass = PyObject_GetAttrString(pyresponse_type, "__class__"); \
+  if (!pymetaclass) { \
+    return PyErr_Format(PyExc_AttributeError, \
```
return `NULL` since `PyObject_GetAttrString` already set an exception

jacobperron replied:
Done 34db026

sloretz commented on rclpy/src/rclpy/_rclpy_action.c:
```diff
@@ -0,0 +1,929 @@
+  } \
+  PyObject * pymetaclass = PyObject_GetAttrString(pymsg_type, "__class__"); \
+  if (!pymetaclass) { \
+    return PyErr_Format(PyExc_AttributeError, \
```
return `NULL` here too

jacobperron replied:
Done 34db026

sloretz commented on rclpy/src/rclpy/_rclpy_action.c:
```diff
@@ -0,0 +1,929 @@
+  } \
+  PyObject * pytaken_msg = convert_to_py(taken_msg); \
+  destroy_ros_message(taken_msg); \
+  if (!pytaken_msg) { \
```
Nit, can `return pytaken_msg;` without checking it

jacobperron replied:
Removed 34db026

jacobperron commented on rclpy/src/rclpy/impl/common.h:
```diff
@@ -0,0 +1,86 @@
+{
+  PyObject * pyattr = PyObject_GetAttrString(pymetaclass, attr);
+  if (!pyattr) {
+    return PyErr_Format(PyExc_AttributeError,
```
Done 34db026

jacobperron commented on rclpy/src/rclpy/impl/common.h:
```diff
@@ -0,0 +1,86 @@
+  // A success returns 0, and a failure returns -1
+  int set_ret = 0;
+  set_ret += PyObject_SetAttrString(pyqos_profile,
+      "depth", PyLong_FromSize_t(qos_profile->depth));
```
Fixed 34db026

jacobperron commented on rclpy/src/rclpy/impl/common.h:
```diff
@@ -0,0 +1,86 @@
+  PyObject * pyqos_profile = PyObject_CallObject(pyqos_policy_class, NULL);
+  Py_DECREF(pyqos_policy_class);
+  if (!pyqos_profile) {
+    return PyErr_Format(PyExc_RuntimeError, "Failed to call object QosProfile");
```
Updated 34db026

jacobperron commented:
@sloretz I've extended the Waitable API so executors can call Future callbacks e405bd9

sloretz commented on rclpy/rclpy/action_client.py:
```diff
@@ -0,0 +1,495 @@
+        return self._status
+
+
+class ActionClient(Waitable):
```
This doesn't need to be done in this PR, or at all. I would recommend making `ActionClient` have a `Waitable` rather than inherit from it so `help(instance_of_action_client)` doesn't show the methods from the `Waitable` interface.

jacobperron replied:
Makes sense. What might be nicer is changing the `Waitable` API so that the methods have leading underscores. This should hide the `Waitable` methods and avoid having to create a subclass `ActionClientWaitable` for `ActionClient` to instantiate.

jacobperron commented on rclpy/src/rclpy/_rclpy_action.c:
```diff
@@ -0,0 +1,939 @@
+  if (RCL_RET_OK != ret) {
+    PyErr_Format(PyExc_RuntimeError,
+      "Failed to get number of entities for '%s': %s",
+      PyCapsule_GetName(pyentity),
```
In this path of execution, `pyentity` has already been confirmed to be a capsule with the name `"rcl_action_client_t"` or `"rcl_action_server_t"`. If it isn't one of those two capsule types, then we return earlier at L210 or L112:

https://github.com/ros2/rclpy/blob/e405bd9c563f82c2021705a7b05711e23a080145/rclpy/src/rclpy/_rclpy_action.c#L206-L213

sloretz replied:
oops

jacobperron commented:
This is ready for another round of review.

CI:
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=6097)](http://ci.ros2.org/job/ci_linux/6097/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=2572)](http://ci.ros2.org/job/ci_linux-aarch64/2572/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=5025)](http://ci.ros2.org/job/ci_osx/5025/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=5965)](http://ci.ros2.org/job/ci_windows/5965/)

sloretz commented:
CI
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=6104)](http://ci.ros2.org/job/ci_linux/6104/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=2579)](http://ci.ros2.org/job/ci_linux-aarch64/2579/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=5032)](http://ci.ros2.org/job/ci_osx/5032/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=5973)](http://ci.ros2.org/job/ci_windows/5973/)

sloretz commented on rclpy/src/rclpy/impl/common.h:
```diff
@@ -0,0 +1,125 @@
+
+  PyObject * pyqos_policy_class = PyObject_GetAttrString(pyqos_module, "QoSProfile");
+  if (!pyqos_policy_class) {
+    return NULL;
```
decref `pyqos_module` before returning

jacobperron replied:
Fixed 52afeca

sloretz commented on rclpy/src/rclpy/impl/common.h:
```diff
@@ -0,0 +1,125 @@
+  PyObject * pyqos_profile = PyObject_CallObject(pyqos_policy_class, NULL);
+  Py_DECREF(pyqos_policy_class);
+  if (!pyqos_profile) {
+    return NULL;
```
decref pyqos_module before returning, or do it once after getting `pyqos_policy_class` since it isn't used below

jacobperron replied:
Fixed 52afeca

sloretz commented on rclpy/src/rclpy/impl/common.h:
```diff
@@ -0,0 +1,125 @@
+      "avoid_ros_namespace_conventions", pyqos_avoid_ros_namespace_conventions);
+
+  if (0 != set_result) {
+    Py_DECREF(pyqos_depth);
```
I think these must always be `DECREF`'d after calling `PyObject_SetAttrString` because it [`INCREF`'s internally](https://github.com/python/cpython/blob/bb86bf4c4eaa30b1f5192dab9f389ce0bb61114d/Objects/dictobject.c#L1029) rather than stealing a reference.

jacobperron replied:
Fixed 52afeca

jacobperron commented on rclpy/rclpy/action_client.py:
```diff
@@ -0,0 +1,498 @@
+        for seq, req_future in pending_requests.items():
+            if future == req_future:
+                try:
+                    del pending_requests[seq]
```
No explicit tests, but I tried a test sending multiple goals concurrently and still works. It would raise an exception, but the iteration stops on the first deletion because of the return statement below. Anyways, it's probably  safer to cast the items to a list as suggested in the SO post.

We might consider doing the same for the service client code as well:

https://github.com/ros2/rclpy/blob/0fafaec6b3764654413d0ce37839d7c7178245af/rclpy/rclpy/client.py#L70-L76

jacobperron replied:
Cast to list: 04337c9

jacobperron commented on rclpy/rclpy/action.py:
```diff
@@ -0,0 +1,15 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from rclpy.action_client import ActionClient  # noqa
```
Done 7cfb217

jacobperron commented on rclpy/rclpy/action_client.py:
```diff
@@ -0,0 +1,498 @@
+
+from rclpy.impl.implementation_singleton import rclpy_action_implementation as _rclpy_action
+# TODO(jacobperron): Move check_for_type_support to its own module (e.g. type_support)
+#                    Do after Crystal patch release since this breaks API
```
Done 4505849

jacobperron commented on rclpy/rclpy/action_client.py:
```diff
@@ -0,0 +1,498 @@
+            data['status'] = _rclpy_action.rclpy_action_take_status(
+                self._client_handle, self._action_type.GoalStatusMessage)
+
+        if not any(data):
```
Done 04337c9

jacobperron commented on rclpy/rclpy/action_client.py:
```diff
@@ -0,0 +1,498 @@
+            goal_uuid = uuid.UUID(bytes=bytes(feedback_msg.action_goal_id.uuid))
+            # Call a registered callback if there is one
+            if goal_uuid in self._feedback_callbacks:
+                self._feedback_callbacks[goal_uuid](feedback_msg)
```
Done 04337c9

jacobperron commented on rclpy/rclpy/action_client.py:
```diff
@@ -0,0 +1,498 @@
+        self._sequence_number_to_goal_id = {}  # goal request sequence number
+        self._pending_cancel_requests = {}
+        self._pending_result_requests = {}
+        self._feedback_callbacks = {}
```
Done 04337c9

jacobperron commented on rclpy/rclpy/action_client.py:
```diff
@@ -0,0 +1,498 @@
+            num_entities[1],
+            num_entities[2],
+            num_entities[3],
+            num_entities[4])
```
Done 04337c9

jacobperron commented:
@sloretz I've addressed all of comments.

I've hidden the `cancel_goal` and `get_result` calls in `ActionClient` and added facades to `GoalHandle` (12907d7). It seemed like the easiest route considering the `ActionClient` is needed for making the outgoing/incoming requests.

Perhaps if we add the notion of a client goal handle to `rcl` and give it (and the server goal handle) more responsibility related sending/receiving cancel and result requests it will tie in more naturally at the `rclpy` and `rclcpp` layers. This is just a thought and outside the scope of this PR of course.

sloretz commented on rclpy/rclpy/action/client.py:
```diff
@@ -34,17 +33,18 @@
-            raise TypeError('Expected UUID, but given {}'.format(type(goal_id)))
-
+    def __init__(self, action_server, goal_id, goal_response):
+        self._action_server = action_server
```
`server` -> `client`?

jacobperron replied:
Yeah, oops. Context switching from working on action server :smile:

jacobperron replied:
Fixed: 102de88

sloretz commented on rclpy/rclpy/action/client.py:
```diff
@@ -120,11 +158,18 @@ def __init__(
 
         self._is_ready = False
+
+        # key: UUID in bytes, value: ClientGoalHandle
```
Recommend `value` being a [weak reference](https://docs.python.org/3/library/weakref.html) to a void a reference cycle

jacobperron replied:
Now weak references: e2df735

jacobperron commented:
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=6110)](http://ci.ros2.org/job/ci_linux/6110/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=2585)](http://ci.ros2.org/job/ci_linux-aarch64/2585/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=5039)](http://ci.ros2.org/job/ci_osx/5039/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=5980)](http://ci.ros2.org/job/ci_windows/5980/)

jacobperron commented:
Increased test timeout:

* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=6113)](http://ci.ros2.org/job/ci_linux/6113/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=2588)](http://ci.ros2.org/job/ci_linux-aarch64/2588/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=5042)](http://ci.ros2.org/job/ci_osx/5042/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=5983)](http://ci.ros2.org/job/ci_windows/5983/)

jacobperron commented:
I'm not sure why, but the multi-threaded test was timing out on CI machines. I've disabled it and opened an issue (#268) to come back to.

* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=6117)](http://ci.ros2.org/job/ci_linux/6117/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=2590)](http://ci.ros2.org/job/ci_linux-aarch64/2590/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=5044)](http://ci.ros2.org/job/ci_osx/5044/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=5985)](http://ci.ros2.org/job/ci_windows/5985/)
